### PR TITLE
Enhancement: Reading & analysing mailing lists in the LISTSERV16.5 format

### PR DESCRIPTION
**Pull request by Christovis**
This PR contains a suggestion to extend BigBang to read and analyze mailing list in the LISTSERV16.5 format. Examples that use this format are [3GPP](https://list.etsi.org/scripts/wa.exe?INDEX) and IEEE. This is done by converting the files from  LISTSERV16.5 to GNU-Mailman and use pythons standard library mailbox to do the rest (thus building on top of analysis routines that are already present in BigBang).

Notable Changes:
The Archive class has now some classmethods to initialize it from one(multiple) file(s) or a dataframe. Below is an example to initialize it from multiple files with one directory.
```
Archive.from_files(
    dir_paths=["../../archives/3GPP_TSG_SA_WG2_UPCON/"],
    file_names=["3GPP_TSG_SA_WG2_UPCON*"],
    email_list_software="LISTSERV",
)
```

Pleas note also that this PR contains the pre-commit routine I suggested in this [PR](https://github.com/datactive/bigbang/pull/403). Therefore there are also changes in other files.

### CODE REVIEW

**Comment by sbenthall** [comment]
Let's merge #403 then come back to this.
It will be easier to see the specific changes for this PR then I think.
The thing I can't identify yet without a deeper review is --- are there automated tests for the new functionality?
Ideally, some dummy data in LISTSERV format. It's a bit exotic, so I wouldn't even know what that looks like.

**Comment by Christovis** [comment]
Yep, I totally agree :+1: 
I was just too impatient and wanted show Niels how things are atm.  :-P
Tests for the new listserv.py file are added now.

File: bigbang/archive.py [hunk 1]
```diff
@@ -24,89 +33,198 @@ class Archive(object):
...
         Upon initialization, the Archive object drops duplicate entries
         and sorts its member variable *data* by Date.
         """
-          
-        if isinstance(data, pd.core.frame.DataFrame):
-            self.data = data.copy()
-        elif isinstance(data, str):
-            self.data = mailman.load_data(data,archive_dir=archive_dir,mbox=mbox)
-        
+        self.data = df
```
Comment (sbenthall) [inline bigbang/archive.py:53]: What happened to the string input case?
The old version copies the data whereas the new version reuses the old dataframe.
So if some other process mutates the dataframe, the new archive is effected.

Why did you make these changes?

File: examples/name-and-gender/senders_guessed_gender-3GPP_TSG_SA_WG2_UPCON.csv [hunk 2]
```diff
@@ -0,0 +1,60 @@
+From,Message Count,Name,First Name,Guessed Gender
+Ali Mouafik  <b'aliamine.mouafik@ORANGE.COM'>,4.0,Ali Mouafik ,Ali,unknown
+Alla Goldner  <b'agoldner@ALLOT.COM'>,110.0,Alla Goldner ,Alla,female
```
Comment (sbenthall) [inline examples/name-and-gender/senders_guessed_gender-3GPP_TSG_SA_WG2_UPCON.csv:3]: I don't think it's a good idea to include this list of people's names, emails, and guessed genders in the repository.
It is a small privacy red flag (even though it is derived from public data....)

Is this data needed for testing, or is it generated by the code?

File: tests/data/3GPP/3GPP_TSG_SA_WG2_UPCON.LOG1308A [hunk 3]
```diff
@@ -0,0 +1,3639 @@
...
+<meta name=3D"Generator" content=3D"Microsoft Exchange Server">
+<!-- converted from rtf -->
+<style><!-- .EmailQuote { margin-left: 1pt; padding-left: 4pt; border-left:=
+ #800000 2px solid; } --></style>
+</head>
+<body>
+<font face=3D"Calibri" size=3D"2"><span style=3D"font-size:11pt;">
+<div>Dear all,</div>
+<div>&nbsp;</div>
+<div>&nbsp;</div>
```
Comment (sbenthall) [inline tests/data/3GPP/3GPP_TSG_SA_WG2_UPCON.LOG1308A:58]: Is it legal for us to include this data? It looks like it is from 3GPP lists. Are there any licensing issues thare? @nllz

**Comment by sbenthall** [comment]
Please see comments on the code.
Also, the code fomatting PR #408 looks like it has conflicts with this PR? please resolve merge conflicts
when that's done I'll test locally

File: examples/name-and-gender/senders_guessed_gender-3GPP_TSG_SA_WG2_UPCON.csv (hunk 2 above)
Comment (Christovis) [reply examples/name-and-gender/senders_guessed_gender-3GPP_TSG_SA_WG2_UPCON.csv:3]: Yes you are right!
I was constructing this list routines on the archive Niels provided us a link to.  Thus for this part, data needed to be available locally.
But it is neither needed for the tests or the code as soon as the code that can scrape the mailing list is implemented (which is ready on my repo).

**Comment by Christovis** [comment]
As I got a working the listserv 16.5 scraper on my local repo, which avoids needing to have a data containing peoples names, email addresses, and message content for test purposes.
Therefore I think it is best to cancel this PR.
I hope to open a new PR asap in which I have a unified routine to scrape, read and save various data formats of listserv 16.5 mailing lists.

File: bigbang/archive.py (hunk 1 above)
Comment (Christovis) [reply bigbang/archive.py:53]: I have implemented various classmethods for the initialization of the class. This requires the user to now from what data type they want to create the class (e.g. url, string, dataframe). The case for loading the archive from a file is outlined [here](https://github.com/Christovis/bigbang/blob/c47b8ccf0cd9403899269832362a43353c222cc9/bigbang/archive.py#L145). 

It seems to me that this approach is more common now-a-days as outline [here](https://www.programiz.com/python-programming/methods/built-in/classmethod).

We can of course implement a classmethod, from_any(), that initialization the class and assumes now knowledge of what is put in (thus url, string, or dataframe have to distinguished).

**Comment by sbenthall** [comment]
Ok, on your recommendation I will close this PR.

I look forward to reviewing the new one when it's ready! This is an exciting feature and I'm sure it will lead to some great work going forward.

**Pull request closed without merging**